### DISCUSSION TITLE
Opt out of scrolling on hash change

### DISCUSSION THREAD

**Discussion by elliottgrieco-toast** [TanStack/router#1155] [Ideas]
I would like to opt out of scrolling in hash change. I am working with a code base which has a feature which controls the hash history external based upon the scrolling position. Having an effect which responds to the history and scrolls to the dom element is a sensible default, but would like to opt out of possible!

PR already created - https://github.com/TanStack/router/pull/1105 , looking for discussion if this feature would be valuable to others. Right now I'm handling this in my project by rewriting the entire RouterProvider, which is not ideal.

**Comment by SeanCassiere** [comment]
Hey, I dropped a [comment](https://github.com/TanStack/router/pull/1105#issuecomment-2019026150) on your PR.

**Comment by elliottgrieco-toast** [comment]
Hi all, we are looking for some more concrete use-cases where this feature/API would solve a problem for developers. Are there cases for folks where the default assumption to scroll in all history hash changes is problematic for people's real world applications? If so, how?

**Comment by joshuaslate** [comment]
I am running into a similar issue.

I am trying to implement a table of contents sidebar for an API documentation site that contains anchor links (e.g., to="#Auth") to the various sections of documentation (all in the same page). When clicking on one of the table of contents links, having the browser scroll down to the anchor that matches the hash is working as expected. The part I'm running into some trouble with is when I try to update the hash as a result of the user scrolling down the content part of the site. When I try to update the hash in the URL, the browser scrolls to the anchor element, but I'd really like to avoid that behavior so the user can scroll at their own pace. I've tried using `navigate`, updating the location state using `router.buildAndCommitLocation`, and even just trying `window.location.hash = '#Auth'`. All of these result in the hash being updated in the address bar, but they also all take over scrolling. I believe the latter is happening because the history implementation is overwriting `window.history` to support the router's subscriptions.

Since the original PR was created, it seems the code causing the scroll has moved here: https://github.com/TanStack/router/blob/35af575ab4c623556ecdb613ac1c85864f0c95d9/packages/react-router/src/Transitioner.tsx#L146

If I'm understanding correctly, the recommendation from the discussion in the PR that was closed was to try using the Scroll Restoration API. I wasn't able to get that to work either, unfortunately.

Here's a StackBlitz that hopefully minimally demonstrates the issue: https://stackblitz.com/edit/tanstack-router-785cuome?file=src%2Fmain.tsx I also tried using `useElementScrollRestoration`, but it didn't solve the issue either. (The intersection observer implementation here isn't without issue, but it should be fine to demonstrate the problem I have, I think).

**Comment by satoshionoda** [comment]
I believe I'm encountering the same issue as @joshuaslate, even the use case is identical, an API documentation site with a sidebar index.

When scrolling, the API documentation (which is generated by a third-party program) replaces the browser hash, and TanStack Router responds to these events, causing annoying jumps.

After digging through the TanStack Router codebase, I discovered that it replaces window.history.replaceState: [View source](https://github.com/TanStack/router/blob/main/packages/history/src/index.ts#L533).

To work around this, I restored the default browser behavior for replaceState as follows. Although this solution might have some drawbacks, it has worked well for my use case so far:

```TypeScript
const originalReplaceState = window.history.replaceState;
const router = createRouter({
  YOUR_OPTION_HERE
});
window.history.replaceState = originalReplaceState;
```

But anyway this is quite hacky and it would be great TanStack Router has its own API for ignoring hash changes.

**Comment by flipvh** [comment]
Hi @elliottgrieco-toast I have been interested in doing something like the scalar docs are doing: fluidly update the hash when scrolling. Here is an example: https://api.cellajs.com/docs . I haven't figured out yet how to do this with React + tanstack router, but perhaps I am missing things. Could this be a usecase for what you are proposing?

**Comment by tpxxx** [comment]
In my case, updating the browser's hash with hashScrollIntoView set to false solved the issue:
`navigate({ hash:newHash, hashScrollIntoView: false });`

**Comment by tervay** [comment]
This can happen when using components like Scalar API docs; in that case you can manually disable scroll restoration on specific routes:

```ts
const router = createRouter({
  // ...
  scrollRestoration: ({ location }) => {
    return location.pathname !== '/apidocs';
  },
  // ...
});
```
